src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -0,0 +1,1771 @@
+        _ = a is [0] and {1};         // 2
+        _ = a is [1] and {1};
+        _ = a is {Length:0} and {1};  // 3
+        _ = a is {Length:1} and {1};
+        _ = a is {1,2,3} and {1,2,4}; // 4
+        _ = a is {1,2,3} and {1,2,3};
+        _ = a is ([>0]) and ([<0]);   // 5
+        _ = a is ([>0]) and ([>=0]);
+        // PROTOTYPE(list-patterns) Parsing length patterns inside combinators
+    } 
+}
+";
+            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithListPatterns);
+            compilation.VerifyEmitDiagnostics(
+                // (7,13): error CS8518: An expression of type 'int[]' can never match the provided pattern.
+                //         _ = a is [0]{1};              // 1
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "a is [0]{1}").WithArguments("int[]").WithLocation(7, 13),
+                // (9,13): error CS8518: An expression of type 'int[]' can never match the provided pattern.
+                //         _ = a is [0] and {1};         // 2
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "a is [0] and {1}").WithArguments("int[]").WithLocation(9, 13),
+                // (11,13): error CS8518: An expression of type 'int[]' can never match the provided pattern.
+                //         _ = a is {Length:0} and {1};  // 3
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "a is {Length:0} and {1}").WithArguments("int[]").WithLocation(11, 13),
+                // (13,13): error CS8518: An expression of type 'int[]' can never match the provided pattern.
+                //         _ = a is {1,2,3} and {1,2,4}; // 4
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "a is {1,2,3} and {1,2,4}").WithArguments("int[]").WithLocation(13, 13),
+                // (15,13): error CS8518: An expression of type 'int[]' can never match the provided pattern.
+                //         _ = a is ([>0]) and ([<0]);   // 5
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "a is ([>0]) and ([<0])").WithArguments("int[]").WithLocation(15, 13)
+                );
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void ListPattern_Interface()
+        {
+            var source = @"
+D.M(new C());
+
+interface I
+{
+    int Length => 1;
+    int this[int i] => 42;
+    string Slice(int i, int j) => ""slice"";
+}
+class C : I
+{
+}
+class D
+{
+    public static void M<T>(T t) where T : I
+    {
+        if (t is not [var length] { var item, ..var rest }) return;
+        System.Console.WriteLine(length);
+        System.Console.WriteLine(item);
+        System.Console.WriteLine(rest);
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition },
+                targetFramework: TargetFramework.NetCoreApp,
+                parseOptions: TestOptions.RegularWithListPatterns);
+            Assert.True(comp.Assembly.RuntimeSupportsDefaultInterfaceImplementation);
```

Comment:
nit: instead of conditioning the whole test, if think you could do `VerifyEmitDiagnostics(); if (RuntimeSupportsDefaultInterfaceImplementation) { CompileAndVerify }; `

Reply:
Could `C` implement `I` directly, to avoid using default interface methods, allowing the test to run on all platforms?

Reply:
That would work too :-)